### PR TITLE
[DRAFT] MultiDB support

### PR DESCRIPTION
**Pull request by flea89**
This PR is to add support to mass-migration to multiple DBs.

### Assumptions

- There's a requirement to be able to force a migration to run on specific databases
- By default, the migration can run on any DB
- There's always the option to run the migration letting django decide which database to use (so even when `allowed_db_aliases = []` the migration can still be run).
- Migration records might be stored on a different DB from the migration one.
- Migration records are stored on the default database by default.
- For the first iteration of this, all the different migrations against different dbs are all listed in the `manage` view.

### Todo (in this ticket):
- Update documentation to explain how this all works
- Find a better name for `_get_queryset_without_namespace`

###  Notes:
- This is a breaking change.
- This currently doesn't work because of [this bug](https://gitlab.com/potato-oss/djangae/djangae/-/issues/1368) in djangae. defer_with_finalize always use the default DB.

###  Improvements (different tickets):
- Allow to pick a specific DB from `manage` to filter only those migrations.

### CODE REVIEW

File: massmigration/migrations.py [hunk 1]
```diff
@@ -7,18 +8,31 @@
...
 
 
+def _get_valid_db_aliases():
+    # We append None here to allow the migration to run without specifying a using and therefore
+    # allowing django/django router to do its course.
+    return list(settings.DATABASES.keys()) + [None]
+
+
+def _is_valid_db_alias(db_alias):
+    if db_alias in _get_valid_db_aliases():
```
Comment (adamalton) [inline massmigration/migrations.py:31]: This could be reduced to just `return db_alias in _get_valid_db_aliases()`.

File: massmigration/migrations.py [hunk 2]
```diff
@@ -33,14 +47,57 @@ class BaseMigration:
...
+        Returns a list of allowed database aliases for running the
+        migration assuming that letting Django (and Django router) do its
+        course is always an allowed option.
+        If `allowed_database_aliases` is None, the migration can be run on
+        all databases.
+        """
+
+        if cls.allowed_database_aliases is None:  # If None we assume can run on all DBs
+            available_dbs = _get_valid_db_aliases()
+            if len(available_dbs) > 1:  # If we only have one DB there's no need to add it, since we alread have the None
```
Comment (adamalton) [inline massmigration/migrations.py:68]: If I'm reading the code correctly, then if this `if` condition is false, we'll get an undefined variable at the `return` statement?

File: massmigration/migrations.py [hunk 3]
```diff
@@ -33,14 +47,57 @@ class BaseMigration:
...
+        else:
+            allowed_db_aliases = cls.allowed_database_aliases + [None]
+
+        return allowed_db_aliases
+
+    def __init__(self, app_label, name, database_alias=None):
         self.app_label = app_label
         self.name = name
+        self.database_alias = database_alias
+        if not _is_valid_db_alias(self.database_alias):
```
Comment (adamalton) [inline massmigration/migrations.py:79]: It looks like this logic is a duplicate of what's in `_check_allowed_db_alias`?

**Review by adamalton — COMMENTED** [review]
So (if I'm reading the code changes correctly) it looks like this implementation will allow each migration to run on _one_ of the databases. Whereas I think what we want is to allow each migration to run on _one or more_ databases. E.g. if we have a migration for `0001_add_default_value_to_users_table` and our users table exists in more than one database (because we're regionalising users into geo-located DBs based on their location) then we should be able to run that same migration on all of our DBs.

In other words I think we should probably have a `using` kwarg in the `BaseMigration.launch` method, so that the same migration can be launched on each of the (allowed) DBs.

So then in the 'manage' page we'd have something that looks something along the lines of this:
Screenshot 2024-03-04 at 16 46 05
Does that seem like the right approach? Am I right in saying that that's not possible in the current approach? What do you think?

**Comment by flea89** [comment]
Closing in favour of #2
